src/grammar/trigger.rs

```diff
@@ -0,0 +1,643 @@
+// SPDX-License-Identifier: SEE LICENSE IN LICENSE.md
+// SPDX-FileCopyrightText: 2023 CYBERTEC PostgreSQL International GmbH
+// <office@cybertec.at>
+
+use crate::lexer::TokenKind;
+use crate::parser::{expect_token_match, Parser};
+use crate::syntax::SyntaxKind;
+
+use super::*;
+
+/// Parses a complete trigger.
+pub(crate) fn parse_trigger(p: &mut Parser) {
+    p.start(SyntaxKind::Trigger);
+    parse_header(p);
+    parse_body(p);
+    p.finish();
+}
+
+/// Parses the header of a trigger.
+fn parse_header(p: &mut Parser) {
+    p.start(SyntaxKind::TriggerHeader);
+    p.expect(T![create]);
+    if p.eat(T![or]) {
+        p.expect(T![replace]);
+    }
+
+    p.eat_one_of(&[T![editionable], T![noneditionable]]);
+
+    p.expect(T![trigger]);
+    parse_ident(p, 1..2);
+
+    expect_token_match!(
+        p,
+        &[TokenKind::BeforeKw] | &[TokenKind::InsteadKw, TokenKind::OfKw] | &[TokenKind::AfterKw] =>
+            match p.current() {
+                T![insert] | T![update] | T![delete] => parse_simple_dml_trigger(p),
+                _ => parse_system_trigger(p)
+            },
+        &[TokenKind::ForKw] => p.error(ParseError::Unimplemented("compound trigger".to_string()))
+    );
+
+    p.finish();
+}
+
+fn parse_simple_dml_trigger(p: &mut Parser) {
+    parse_dml_event_clause(p);
+    parse_referencing_clause(p);
+
+    if p.eat(T![for]) {
+        p.expect(T![each]);
+        p.expect(T![row]);
+    }
+
+    parse_trigger_edition_clause(p);
+    parse_trigger_ordering_clause(p);
+
+    p.eat_one_of(&[T![enable], T![disable]]);
+
+    if p.eat(T![when]) {
+        p.expect(T!["("]);
+        parse_expr(p);
+        p.expect(T![")"]);
+    }
+}
+
+fn parse_system_trigger(p: &mut Parser) {
+    loop {
```

Comment:
After doing some inspection, I think it's possible to rewrite the `match` statement in a way that the result is concise, intentional & readable. When using the `T!` macro with the `Parser::lookahead` function this loop could look like:

```rust
loop {
    match p.lookahead(3).as_slice() {
        [T![alter], ..] => (),
        [T![analyze], ..] => (),
        [T![audit], ..] => (),
        [T![comment], ..] => (),
        [T![create], ..] => (),
        [T![ddl], ..] => (),
        [T![drop], ..] => (),
        [T![grant], ..] => (),
        [T![noaudit], ..] => (),
        [T![rename], ..] => (),
        [T![revoke], ..] => (),
        [T![truncate], ..] => (),
        [T![associate], T![statistics], ..] => (),
        [T![disassociate], T![statistics], ..] => (),
        [T![after], T![clone], ..] => (),
        [T![after], T![db_role_change], ..] => (),
        [T![after], T![logon], ..] => (),
        [T![after], T![servererror], ..] => (),
        [T![after], T![startup], ..] => (),
        [T![after], T![suspend], ..] => (),
        [T![after], T![set], T![container]] => (),
        [T![before], T![logoff], ..] => (),
        [T![before], T![shutdown], ..] => (),
        [T![before], T![unplug], ..] => (),
        [T![before], T![set], T![container]] => (),
        _ => {
            p.error(ParseError::UnknownToken("Unknown".to_string()));
            return;
        }
    }

    p.bump_any();
    if !p.eat(T![or]) {
        break;
    }
}
```

The `match` arms check all sets of supported sequences, containing 1, 2 or 3 `TokenKind` to check. This way there is no nesting involved, the arms specify the full set of allowed keyword sequences, thereby improving readability. The call to `p.bump_any()` is explicit, so is the error case that logs an error and returns from this function.

What do you think?

Reply:
Haha, I pretty much have the exact same structure in my current working changes locally, albeit that I grouped them depending on their amount of tokens, so that in the match arm `p.bump_any()` is called either once, twice or thrice.

Reply:
:+1:

Realized late that of course the number of bump calls has to match the number of tokens found, for example each arm could return the number to advance (instead of `()`).